### DISCUSSION TITLE
[Discussion] .NET MAUI Single Project

### DISCUSSION THREAD

**Discussion by pauldipietro** [dotnet/maui#70] [General]
# .NET MAUI Single Project

*This is an evolving discussion and will be updated as new questions/subjects arise.*

During our [Build session](https://mybuild.microsoft.com/sessions/1e045a36-71b0-41c4-9679-83abc027d25f?source=sessions) about .NET MAUI we discussed plans to provide a single project structure for developing apps. Presently, Xamarin.Forms apps utilize a project for shared code and a platform-specific project head for each specific target platform supported, e.g.:

- MyApp
- MyApp.Android
- MyApp.iOS
- MyApp.UWP

A couple high level goals of the single project are to make developing apps with .NET MAUI more immediately digestible for those new developers and simplify the current structure for those migrating existing Xamarin.Forms apps, along with reducing the burdens that come with maintaining multiple project heads. We hope this helps provide an amazing experience for developers in the future.

During the session we presented a structure of how a single project might look; we can call this **Grouping by Platform**. We also have a second concept that we'd like to present to the community to receive feedback on, which we'll call **Grouping by Functionality**.

### Grouping by Platform

As discussed during the Build session, grouping by platform uses a structure where all targeted platforms are represented inside of a `Platforms` directory, and within each of those nested platform-specific subdirectry will be all of the necessary functionality. Here is a visualized mockup from the session:

![Group by Platform](https://user-images.githubusercontent.com/1251024/82493438-f649a100-9ab5-11ea-8fbb-da019a9c32d8.PNG)

In this scenario, the `BluetoothAdapter.android.cs`, `BluetoothAdapter.ios.cs`, and `BluetoothAdapter.windows.cs` live inside of the Android, iOS, and Windows subdirectories, respectively.

### Grouping by Functionality

Grouping by functionality is another concept we have discussed and would like to present here for discussion. In this structure, a `Platforms` directory is not used, but rather the specific functionality is contained within the various elements of an app. To better illustrate this is another mockup:

![Group by Functionality](https://user-images.githubusercontent.com/1251024/82493655-4e80a300-9ab6-11ea-89bd-31b4faf3d5b1.png)

In this scenario, our app happens to require a custom control which inherits from a `Button`. Inside of the `Controls` folder we use is a `CustomButton.cs` file; this functions similarly to if you currently defined such a control and its properties in the shared project of your Xamarin.Forms app. In this structure, however, the platform-specific code is nested beneath `CustomButton.cs` using the `CustomButton.platform.cs` naming. 

It should be noted that in this scenario, the `Resources` functionality remains the same.

**What are your thoughts on each of the above?**

### Launch Screens

Both Android and iOS require developers to add a launch screen and Windows requires a splash screen to be added to applications; how could we provide a way to provide a single definition of such a screen in a single project that provides an easy way for developers to create such screens?

### Localization

How can we improve the developer experience with regards to localizing apps? Both text as well as localized images must be considered.

### Platform Configuration

Where would the platform-specific configurations such as `AndroidManifest.xml`, `Info.plist`, or `Package.appxmanifest` reside? Also, how and where might common properties across all three platforms be set?

### Resource Management

If for some reason a user needs a specific image resource for a given platform, where and how would that resource be identified and utilized?

## We need your feedback!

Your feedback is *extremely* important here. What are your expectations for this single project structure? Do you have any concerns in a scenario where you would migrate an existing Xamarin.Forms app? If we missed something, let us know!

**Comment by dhindrik** [comment]
I think it is nice to group by functionality when we are talking about code files. But we still need a platform folder for info.plist, AndroidManifest.etc. The platform folder can also be a good place for platform-specific resources, a CoreML model for iOS for example.

**Comment by charlesroddie** [comment]
I don't understand the motivation here. The current project structure makes sense to me.

**For example:** we create a UWP project. It makes sense by itself and displays a blank window. We edit settings, icons, as normal for UWP projects. We create a netstandard project, put a Xamarin Forms page in, and reference it from the UWP project. It shows up there. Then we customize the UWP titlebar, put back and forward buttons there, all using UWP classes.

How does all this work with a single project shared by many platforms? Do we install uwp/ios/android/mac nuget packages into the same project? Who maintains the uniproject?

Currently there are simple separated platform projects. Putting things in one place wouldn't make things simpler except from the most superficial perspective. It's not as if there would be a reduction in files or settings: exactly the same ones would be needed surely (except for a few shared resources).

**Comment by PureWeen** [comment]
> We edit settings, icons, as normal for UWP projects. 

And then you have to do that all again for Android, all again for iOS, all again for WPF, all again for GTK, all again for macOS, all again for ....

The idea with single project is that you can configure as much as possible once which will use the most rational defaults. 

> How does all this work with a single project shared by many platforms? Do we install uwp/ios/android/mac nuget packages into the same project? Who maintains the uniproject?

There's only one nuget and then in the IDE you can just indicate what platforms you want to support.  The Application class becomes a native implementation so there is only one Application and then when you operate under the context of a specific platform you get those benefits.

If as a user you want to add UWP you don't have to go look up in our docs about adding a uwp project, changing a base class, installing nugets, etc...

You just check a box that says.. Hey also build UWP please.

With single project you get all the things you mentioned in your comment but things are just arranged better, and more organized. 

The majority of modern 3rd party library has moved to multi targeting using SDK Extras because it's just easier.

**Comment by ghuntley** [comment]
For code, if not using shared projects, my preference is grouping by functionality with per platform icons in the solution window implemented by a naming convention — ClassName.TFM.cs — that's included via a globbing pattern. No special or extra csproj metadata.

1) Can you please flesh out in this GitHub issue what the actual raw csproj implementation could potentially look like? It's material / would flip folks (inc. me) opinions as to the path chosen.

2) What is the future of shared projects and have they been considered? @jeromelaban can speak mountains to their usefulness/use cases. `partial` classes are a solution here that significantly reduces codebase complexity.

**Comment by ChaseFlorell** [comment]
I love the idea of Grouping by Functionality, however it's not as practical in the scenario where one platform  will require a number of supporting classes to achieve the functionality required.

I think a hybrid approach would be fantastic, whereby a `partial` class with the suggested naming convention would show the file linked under the "functionality" class (option 2), but also be visible under the "platform" section (option 1).

Also, I second @ghuntley's sentiment about shared projects 😄 they are definitely useful.

**Comment by coman3** [comment]
I personally think that a **mix of both** grouping concepts may actually be the best solution here.
##### What: 
- Grouping by functionality should be used for controls / programmatic logic (.cs files in general really).
- Grouping by platform should be used for assets and platform specific configuration.
##### Why:
- We are used to our code being placed in the same 'file' or 'group' for different functionality in other .NET projects. Eg. App.xaml -> App.xaml.cs
- It is a requirement to separate assets and other configuration for each platform. Eg. Android xml docs, iOS plist's,

**Comment by kevinjpetersen** [comment]
Amazing to see how open this project is! I'll be sure to help out and give feedback where I can, as I've definitely been waiting for something like this for a long time! Long time Xamarin.Forms user here :)

Here are some of my thoughts on all of these things.

**Grouping Concepts**
For the Grouping Concepts, I'd personally prefer the Grouping by Platform, so I can tuck away the platform specific classes when I don't need it, and just focus on the shared code in the MyApp and not get distracted by other files that isn't used on a shared level.

**Launch Screens**
I definitely see a world where you can choose to use a single launch screen that works and looks the same for all platforms. The concept of "Write once, run everywhere" should be applied as much as possible where it makes sense, and here it definitely does.

Altho, there should still be a possibility to choose a "Custom" option if you want to createa launch screen for each platform individually.

**Localization**
I like the way Images and Fonts are represented in the above Grouping images, and I think having a way to reuse those for localization would be good. Imagine right-clicking on an Image and having a Property for which language that specific image should apply to. Then showing a small Country Icon or Language Code next to the file name similar to how in the Grouping By Functionality- image, there is an Android, iOS and Windows logo.

**Platform Configuration**
I think having the ability to still access the platform specific files is key to any development process, so maybe keep them in a structure similar to Grouping by Platform- image, where you can access them from that list.

In terms of common or shared properties (Like App name, Package Name, Version, etc.), that might make sense to keep in the main info of the project itself (when you right-click the project to see it's properties). Just thinking out loud here, maybe there's another way to do that.

**Resource Management**
For resource retrieval, both at runtime and before compilation, I think having a unified shared API in XAML and C# to access the Images is the (in my opinion) the best way to keep everything simple and easy. However there should still be the possibilities to dig into a specific platform if you're set on getting the path for that specific platform.

Idea of how that could look:
```csharp
MAUI.Resources.Images.MyCoolCatImage
MAUI.iOS.Resources.Images.MyCoolCatImage
MAUI.Android.Resources.Images.MyCoolCatImage
```

I'm really excited for MAUI, and I hope these thoughts help a little bit :)

**Comment by Happypig375** [comment]
>Resource Management
For resource retrieval, both at runtime and before compilation, I think having a unified shared API in XAML and C# to access the Images is the (in my opinion) the best way to keep everything simple and easy.

How does this compare to embedded resources?

**Comment by JohnH-FF** [comment]
Hi @PureWeen,
> > We edit settings, icons, as normal for UWP projects.
> 
> And then you have to do that all again for Android, all again for iOS, all again for WPF, all again for GTK, all again for macOS, all again for ....

All of those platforms have different icon requirements, so you are going to have to do 'all again' for each platform you want to support anyway, right? Just in a single definition project rather than platform specific? I don't see any advantage in that.

We currently target 4 platforms (ok not quite MacOS but we prototyped a solution for that). We have a separate startup project for each platform, and set up the launch icons and splash screens in each because they are platform specific. How does this suggestion improve that?

Seperately, I DO see a massive advantage of having fonts etc in a shared project. But for shared images there are still the platform differences, x1 x2 x3 for iOS, dpi's for Droid etc. How will you address those?

**Comment by charlesroddie** [comment]
@PureWeen Thanks for clarifying the purpose of single projects. I think we should measure how much sharing there is to assess the benefits of single-project for app developers.

**Fonts:** I think this is covered already in the current Xamarin.Forms.

**Images:** Yes currently annoying since overuse of bitmaps creates more images than needed per platform, and there is extra work to adjust fullscreen/orientation etc.. If would be conveninent if we could specify a few svgs in a single project and generate all the bitmaps and create the appropriate Android `Activity` etc. automatically. If not, then as @JohnH-FF says there is no benefit of single project for images.

**Settings:** I don't see a lot of commonality between settings across uwp/android/ios/mac currently so don't see a large benefit from single-project here. (Edit: post below has verison numbers. That seems useful to make common.)

**Comment by felipebaltazar** [comment]
I think that "settings solution" will be creating another way to group common properties (like permissions, version and etc) and give some editable "file" to specific properties.

Maybe `AppSettings.json` (like we have on aspnetcore) can be a solution for this:

eg:

```json

{
  "global": {
    "version":1,
    "version-name":"1.0.0",
    "permissions":[
          "use-gps"
     ]
  },
  "android": {
    "install-location": "auto",
    "some-platform-specific": "somevalue"
  },
  "ios": {
    "some-platform-specific": "somevalue"
  }
}
```

Another way is using the [hosting/dependency injection](https://github.com/dotnet/maui/issues/24) such as aspnet core (yeah, too)

eg

```c#
Host.CreateDefaultBuilder()
    .Android(settings =>
    {
        settings.Version = 1,
        settings.VersionName = "1.0.0",
        settings.SomePlatformSpecific = "somevalue"
     })
    .Ios(settings =>
    {
        settings.Version = 1,
        settings.VersionName = "1.0.0",
        settings.SomePlatformSpecific = "somevalue"
     })

....

```

**Comment by kevinjpetersen** [comment]
I’ve seen a few people talking about the generally point and why the need of having a single project instead of the normal 4.

These are my personally opinions.

For once, there is tremendous value in ease and simplicity having to just worry about ONE project over 4. I understand the concerns about different image formats, sizes, etc. but we’re in 2020 so I honestly don’t think that’s a technical challenge we can’t overcome. Now a days, anything is possible.

Imagine if Unity had a separate project for each platform. Wouldn’t you be mad AF? Yes, yes you would 😁

**Comment by PureWeen** [comment]
What @felipebaltazar said!!

Let's see android I have to annotate with these attributes or update this manifest, UWP has these checkboxes, iOS has these plist things that I have to add these various settings to.

How much of this can we normalize? Let's look at deeplinking

What if we could just provide one setting for deeplinking and then that auto sets the plist/intent/uwp things needed for that.

How much can we normalize the settings so users can focus on a common API to achieve what they need to?

**Comment by VladislavAntonyuk** [comment]
I am voting for Grouping by Platform:
for example I have a custom control CustomButton. I need a renderer only for WPF project.
if we Grouping by Platform we add renderer to the specific platform folder and I will have no hesitation that all platforms have necessary implementation:
-platforms
--android
--ios
--wpf
---customRenderer

if we choose Grouping by Functionality, we will see something like this:
-customButton.cs
--customButton.WPF.cs

from this structure I understand that this button will work only for WPF, because I see WPF file. but I originally this may work with all other platforms.

**Comment by bmacombe** [comment]
Seems this is going to be a tooling feature in Visual Studio...why not both with a toggle in the solution explorer?  I could see times both would be useful.

**Comment by m5x** [comment]
Is this issue about tooling feature or real on-disk structure of the project? In following paragraphs I assume it's the latter.

IMHO any structure should be possible but if we are talking about default/recommended structure I vote for grouping by platform. As a good development practice in any MAUI project one should strive to share as much code as possible between the platforms. Platform-specific code should be rare when following this principle and mostly needed only for implementation of general abstractions/services used by shared code. Having these files separated and structured by the platform nicely supplements that principle and promotes its usage.

This approach also solves the problem where to put other platform-specific files - if we already have a directory for each platform these other files can be placed there as well (alongside the code files or to sub-directories). 

For common concepts like splash/launcher screen a tooling would be nice to generate initial versions of required platform-specific files for each platform from common definition.

**Comment by memsranga** [comment]
Grouping by Platform sounds like a better option. We can never get away with the core platform constructs (Activities, Intents, ViewControllers...), and we should not. What if I want to open another Activity for Camera? Sure, out-of-the-box support sounds great, but any sufficiently large application will have to depend on some or the other platform-specific code.

@PureWeen maybe if I missed something but, another problem I see is NuGet targeting. What if I want to add a NuGet the only supports one platform? (like Glide for Android). How will the Nuget Manager know how to check compatibility?

**Comment by VladislavAntonyuk** [comment]
@shanranm 
`
<PackageReference Include="PackageName" Version="1.0.0" TargetPlatforms="net5-anndroid;net5-ios"/>
`

**Comment by Happypig375** [comment]
Or simply a `Conditional` attribute on the `PackageReference`.

**Comment by mgierlasinski** [comment]
I prefer Grouping by Functionality, we can do it already in sdk-style projects with multitargeting. It's much more useful if you have project structured by features and not by layers. Let's say you have directory: `MyProject\Features\Bluetooth`, everything regarding bluetooth is visible when you expand one directory:

`Features`
-`Bluetooth`
--`BluetoothSomething.cs`
---`BluetoothSomething.android.cs`
---`BluetoothSomething.ios.cs`
---`BluetoothSomething.windows.cs`

In grouping by platform you have `Features\Bluetooth` in shared part but you also need copy of this path in each platform directory. For 3 platforms you end up having 4 nested directories to browse when you work with that particular bluetooth feature: 

`Platforms`
-`Android`
--`Features`
---`Bluetooth`
----`BluetoothSomething.android.cs`
-`iOS`
--`Features`
---`Bluetooth`
----`BluetoothSomething.ios.cs`
-`Windows`
--`Features`
---`Bluetooth`
----`BluetoothSomething.windows.cs`

To be honest it's not very different from having that in separate projects, you still have to jump between folders.

**Comment by jcmanke** [comment]
So here's a question. What if I want to keep my Bluetooth code in a separate project from my XAML? I assume I could make another .NET 6 project that has `<TargetFrameworks="net6-android;net6-ios"/>` and add a reference from Bluetooth.csproj to MyApp.UI.csproj and that would probably be okay. But what if I want to stay old-school and keep my existing structure of Bluetooth.csproj, Bluetooth.Android.csproj, and Bluetooth.iOS.csproj? Will this support only adding a project reference for one target framework? What will that look like in the Solution Explorer?

**Comment by kevinjpetersen** [comment]
@jcmanke They already confirmed that MAUI will be backwards compatible with Xamarin.Forms, meaning if you want to, you can still have the 3 different projects like you have now.

However they want to make MAUI more simple and streamline for people, so that people can get started more quickly without having to do a lot of complicated setup stuff and have to worry about multiple projects.

**Comment by Happypig375** [comment]
>They already confirmed that MAUI will be backwards compatible with Xamarin.Forms

Please no! There are already a bunch of issues that greatly enhance the app development experience but are backwards incompatible changes.

**Comment by pauldipietro** [comment]
The plan is for backwards compatibility as @kevinjpetersen mentioned. To that point, if that might lead you to forego using a single project and stick with doing things the same way you do now (potentially even for new projects), please do discuss your reasoning because it's just as valuable for us to know.

**Comment by neptrio** [comment]
Thank you for giving us a voice.
 
@pauldipietro As you asked, I would like to explain why I would do things the same way as I do now.

I like the way my platforms are logically separated under one roof. I have to address a lot of platform specific functionalities. As an example, I need to start a service for Android that I do not need for iOS. I appreciate that I can do this completely independent of the other implementations. 
The same applies to resources. I appreciate being able to share resources across platforms, but I also appreciate being able to keep them clearly separated when I have to comply with platform-specific design guidelines.

When I'm done, I minimize the project and go back to the shared project without having to deal with the platform specific implementations. I find this especially important for larger projects, where my shared project already contains enough files.
Furthermore it gives me security in the build process. I click on the platform specific project and can make settings where I expect them, such as linking-settings or something. As a .NET developer I see this logic in the other frameworks too.

I understand that the market share should be increased and the entry for new developers should be simplified. Not on the project level but in an abstract way compared to Xamarin.Forms this simplification was called Xamarin.Forms-Shell.

Maybe there should be a similar abstraction for MAUI, MAUI-Shell to adress new developers.

For me, I don't see any advantages in restructuring my current or future projects.

**Comment by dansiegel** [comment]
### App Manifests

I think @felipebaltazar is largely on the right track here... though I think a name like `Manifest.json` would be a far better choice than `AppSettings.json` since this is the manifest not settings like here's my backend, etc.

```json
{
  "global": {
    "id": "com.contoso.awesomeapp",
    "displayName": "Awesome App",
    "version": 1,
    "version-name":"1.0.0",
    "permissions":[
          "location"
     ]
  },
  "android": {
    "install-location": "auto",
    "some-platform-specific": "somevalue"
  },
  "ios": {
    "some-platform-specific": "somevalue"
  }
}
```

There are a number of things that could be abstracted rather easily like the app name, and version. Assuming we just kept a name like Manifest.json, we could support localization for certain values like the display name in an Manifest.es.json, etc.

Where I think things get a little more involved is with permissions. While Android for example we might simply add the permission for the Coarse, Fine or Mock GPS location, on iOS we may potentially need to provide 4 different descriptions for why we need the permission. 

There are 3 ways I think this could and should be supported...

1) I'm a big fan of keeping things as simple as possible when possible... so just like we saw above you say you want the `location` permission and you get all of the correct permissions on each platform with some generic default descriptions (that are already localized for us) out of the box for iOS.

2) Instead of simply treating the permissions as an array of string's allow something like the following if someone wants to specify a description:

```json
{
  "permissions":[
    {
      "permission": "location",
      "description": "Big brother is watching..."
    }
  ]
}
```

  **NOTE:** the description would be used for each location description on iOS

3) Allow merging with native manifests under the Platform node. I believe the general rule here is any value that is actually provided in the platform specific manifest should always win over the Manifest.json

![image](https://user-images.githubusercontent.com/3860573/83041890-d0605700-9ff5-11ea-9355-9b9be9641df2.png)

### Images

While I'm very much for simplified Image handling, it is worth remembering that there are times that images may be localized, or platform specific (as I recall the TabbedPage project template for Xamarin.Forms currently has images for the Tabs that are only used on iOS).

This sort of begs the question if we're grouping by feature which I like for things like Fonts/Images (which I wholeheartedly love), do we now introduce a Platforms directory under images if we need something only for iOS/Android.

### Native Bindings

I fully admit for probably 95+% of people this isn't going to ever be an issue... but this would be a potential headache for some developers... There is the occasional case where you need to bind native drivers or other native libraries. 

### Platform Specific Project / NuGet References

On a similar, however much more common and easier to deal with issue are platform specific references say basically anything Android. While MSBuild is able to easily tackle this, and those of us who are already experienced with Multi-Targeting with MSBuild.Sdk.Extras, this isn't even something that we'd blink for... but this is an area in Visual Studio/Visual Studio for Mac that is going to need some love.

**Comment by haideralicspak** [comment]
I have one question in their demo video they showed us demo app on Iphone in which they didn't paired their windows visual studio to mac and ran the app on Iphone from windows visual studio. the thing i want to confirm how did they build the solution for Iphone without mac is it new functionality which i don't think is possible or did i miss the connection part with mac.

**Comment by dansiegel** [comment]
@haideralicspak they were using Hot Restart (not to be confused with XAML Hot Reload). You need to enable it in the Preview Features in Visual Studio and you can use it right now today!

![image](https://user-images.githubusercontent.com/3860573/83164055-9d36ca00-a0c0-11ea-91c9-5f37f4113b5e.png)

**Comment by snickler** [comment]
I would actually like to be able to see MSBuild/CSProj properties that can be applied to the Info.plist and AndroidManifest.xml as build tasks. It would definitely make the experience smoother, especially for cases such as automatically incrementing versions. Having to run an external command or script to pull this off isn't the best.

**Comment by michael-hawker** [comment]
Glad I found this issue! (I had added a comment [over here](https://github.com/microsoft/microsoft-ui-xaml/issues/2555#issuecomment-638681391) in a discussion.) [Also related to this issue in the Reunion repo](https://github.com/microsoft/ProjectReunion/issues/61).

Definitely would prefer the **Grouping by Functionality** as @mgierlasinski called out as well [here](https://github.com/dotnet/maui/issues/70#issuecomment-632764098).

I definitely just see things being more useful to be displayed underneath the file as part of my normal project structure:

```
   📁 Helpers
      🔽 BluetoothAdapter.cs
         - 🤖 Android
         - 🍎 iOS
         - 🐱‍👤 Windows
   🔽 MainPage.xaml
       ▶ MainPage.xaml.cs
```

Clicking on the root file itself would just take you to whatever current platform you have selected for debugging, but you could still drill-in to pull up the alternates (and they'd show up in search of course still).

**Comment by Pinox** [comment]
I would prefer the  **Grouping by Functionality** on all common classes and 
**Grouping by Platform** in non-common classes especially in the config sections of the different platforms where the underlying class names and events are generally different between the platforms.

Please also add Blazor , Winui3.0 Desktop and UWP to the list of supported projects.

**Comment by vdvman1** [comment]
I agree:
- grouping by functionality for common classes that might just have some bits that need to be platform specific
- grouping by platform for things that are completely platform specific like configs

**Comment by Xyncgas** [comment]
MAUI, maybe merge with the blazor project
Currently it supports : 
![image](https://user-images.githubusercontent.com/35312807/207299672-8ddc02d4-8ed0-41ec-9b16-af5574b62711.png)
If it supports browser also, which means I can publish for browser like I can with blazor, would be nice

**Comment by veringm** [comment]
With the single project approach, how do we handle these situations

1. Install a nuget for one platform only
2. Install different versions of a nuget for different platforms

Both situations are supported using the platform-specific project approach in Xamarin.Forms.

**Reply by VladislavAntonyuk** [reply]
use conditions.
```
<ItemGroup Condition="$(TargetFramework.Contains('-android'))">
     <PackageReference Include="Microsoft.Maui.Controls" Version="8.0" />
</ItemGroup>
<ItemGroup Condition="$(TargetFramework.Contains('-ios'))">
     <PackageReference Include="Microsoft.Maui.Controls" Version="9.0" />
</ItemGroup>
```

**Reply by veringm** [reply]
Thanks a lot @VladislavAntonyuk. However, I sense (at the moment at least) that handling dependencies in the platform-specific approach is easier and more clean in situations like this.
